src/datafusion/table_provider.rs

```diff
@@ -0,0 +1,136 @@
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_schema::{Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    common::Statistics,
+    datasource::{
+        listing::{ListingTableUrl, PartitionedFile},
+        physical_plan::FileScanConfig,
+        TableProvider, TableType,
+    },
+    execution::context::SessionState,
+    logical_expr::{Expr, TableProviderFilterPushDown},
+    physical_plan::ExecutionPlan,
+};
+
+use crate::{
+    async_reader::{ZarrPath, ZarrReadAsync},
+    reader::ZarrResult,
+};
+
+use super::scanner::ZarrScan;
+
+pub struct ListingZarrTableOptions {}
+
+impl ListingZarrTableOptions {
+    pub async fn infer_schema(
+        &self,
+        state: &SessionState,
+        table_path: &ListingTableUrl,
+    ) -> ZarrResult<Schema> {
+        let store = state.runtime_env().object_store(table_path)?;
+
+        let zarr_path = ZarrPath::new(store, table_path.prefix().clone());
+        let schema = zarr_path.get_zarr_metadata().await?.arrow_schema()?;
+
+        Ok(schema)
+    }
+}
+
+pub struct ListingZarrTableConfig {
+    /// The inner listing table configuration
+    table_path: ListingTableUrl,
+}
+
+impl ListingZarrTableConfig {
+    /// Create a new ListingZarrTableConfig
+    pub fn new(table_path: ListingTableUrl) -> Self {
+        Self { table_path }
+    }
+}
+
+pub struct ZarrTableProvider {
+    table_schema: Schema,
+    config: ListingZarrTableConfig,
+}
+
+impl ZarrTableProvider {
+    pub fn new(config: ListingZarrTableConfig, table_schema: Schema) -> Self {
+        Self {
+            table_schema,
+            config,
+        }
+    }
+}
+
+#[async_trait]
+impl TableProvider for ZarrTableProvider {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::new(self.table_schema.clone())
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> datafusion::error::Result<Vec<TableProviderFilterPushDown>> {
+        // TODO: which filters can we push down?
+        Ok(filters
+            .iter()
+            .map(|_| TableProviderFilterPushDown::Unsupported)
+            .collect())
+    }
+
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        _filters: &[Expr],
+        limit: Option<usize>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        let object_store_url = self.config.table_path.object_store();
+
+        let pf = PartitionedFile::new(self.config.table_path.prefix().clone(), 0);
+        let file_groups = vec![vec![pf]];
+
+        let file_scan_config = FileScanConfig {
+            object_store_url,
+            file_schema: Arc::new(self.table_schema.clone()), // TODO differentiate between file and table schema
```

Comment:
Hmm, right what's even a "file" here? It occurred to me that running the queries at the "store" level (i.e. a directory with one of more sub-directories with zarr arrays in it) might be a bit hard to handle down the road, but then I don't really see another option. I guess I'll look more into what this does in datafusion, we can decide later.

Regarding partitions, the concept is not too well defined here (compared to e.g. with parquet files), but I actually have an interesting idea for that, that would leverage the fact that zarr arrays typically store rasters, like it's rarely just a list of rows with no organization to it. I'll create a task for that as well later, I'm curious to hear if you think my general idea could be useful.

Reply:
Yeah, the "store" being a physical directory complicates things, but to your point, I think that's the nature of zarr stores 🤷 , but perhaps we'll think of a good way.

Re partitions, I'd certainly be curious to see what you have in mind regarding rasters. At the same time, presuming listing table functionality worked (it currently doesn't w/ stores/directories), I do wonder what a workflow could look like there. E.g.

```
s3://bucket/zarr_store/experiment=123/123.zarr/
s3://bucket/zarr_store/experiment=456/456.zarr/
```

You could then potentially do:

```
SELECT *
FROM zarr_store
WHERE experiment=123
```

or you could query across partitions and datafusion would add the metadata from the partition to the result.

Reply:
Yeah, I mean we can use the classic directory partitioning, that still works, I guess I meant it gets slightly messier with zarr files because of the additional directory structure. But ultimately we should probably support something like that. I'll create an issue to discuss the other thing I had in mind, it's not partitioning per se, but it would be a way to avoid reading potentially a lot of data.